### PR TITLE
[WIP] http: O(1) custom headers

### PR DESCRIPTION
**Pull request by mattklein123**
Implement custom O(1) header registration for header maps.
1) Remove virtual inheritence from header maps.
2) Implement variable inline storage for concrete header maps.
3) Various TODO/logic cleanups from previous header refactors.
4) Demonstrate what custom header registration looks like for the
   CORS filter. More cleanup of this type can be done.

Fixes https://github.com/envoyproxy/envoy/issues/4815

Risk Level: High
Testing: TBD
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
@jmarantz @asraa @adisuissa looking to get some initial feedback on this. I decided to implement this for fun. It was a nice challenge. :)

On one hand, this actually cleans up a bunch of things. On the other hand, it's pretty complicated and is plumbing a bit more of the depths of C++ than I would like. In any case, it works, and I've demonstrated what it looks like to register custom headers from the CORS extension. From the extension perspective, it's trivial to use, so I like that part. If we like this direction I can clean it up ad get all tests passing. There is further work that can be done also around making all of the header add/remove/etc. header matcher stuff be aware of custom headers and use the fast lookup if possible.

File: source/extensions/filters/http/cors/cors_filter.cc [hunk 1]
```diff
@@ -13,6 +13,21 @@ namespace Extensions {
...
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_allow_credentials(Http::Headers::get().AccessControlAllowCredentials);
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_allow_methods(Http::Headers::get().AccessControlAllowMethods);
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_allow_headers(Http::Headers::get().AccessControlAllowHeaders);
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_max_age(Http::Headers::get().AccessControlMaxAge);
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_expose_headers(Http::Headers::get().AccessControlExposeHeaders);
```
Comment (mattklein123) [inline source/extensions/filters/http/cors/cors_filter.cc:29]: This is what it looks like to an extension. Pretty nice I think. :)

File: include/envoy/http/header_map.h [hunk 2]
```diff
@@ -637,12 +631,90 @@ class HeaderMap {
...
+
+  /**
+   * Register an inline header and return a handle for use in inline header calls. Must be called
+   * prior to finalize().
+   */
+  static Handle registerInlineHeader(const LowerCaseString& header_name) {
+    static size_t inline_header_index = 0;
+
+    ASSERT(!mutableFinalized());
+    auto entry = headerNameToIndex().find(header_name);
```
Comment (jmarantz) [inline include/envoy/http/header_map.h:655]: consider capturing the map in a temp so we don't have to call this 4x.

File: include/envoy/http/header_map.h [hunk 3]
```diff
@@ -637,12 +631,90 @@ class HeaderMap {
...
+  using Handle = Phantom<RegistrationMap::const_iterator, T>;
+
+  /**
+   * Register an inline header and return a handle for use in inline header calls. Must be called
+   * prior to finalize().
+   */
+  static Handle registerInlineHeader(const LowerCaseString& header_name) {
+    static size_t inline_header_index = 0;
+
+    ASSERT(!mutableFinalized());
```
Comment (jmarantz) [inline include/envoy/http/header_map.h:654]: it is surprising that you are asserting we are not finalized here, and then in the next line calling an accessor that asserts it is finalized. I don't see the intervening call to finalize().

File: include/envoy/http/header_map.h [hunk 4]
```diff
@@ -651,30 +723,91 @@ class RequestOrResponseHeaderMap : public virtual HeaderMap {
...
+class CustomInlineHeaderUtility {
+public:
+  /**
+   * Finalize all concrete header map registrations.
+   */
+  static void finalize() {
+    CustomInlineHeaderRegistry<RequestHeaderMap>::finalize();
+    CustomInlineHeaderRegistry<RequestTrailerMap>::finalize();
+    CustomInlineHeaderRegistry<ResponseHeaderMap>::finalize();
+    CustomInlineHeaderRegistry<ResponseTrailerMap>::finalize();
```
Comment (jmarantz) [inline include/envoy/http/header_map.h:807]: Ah I see now there are exactly 4 registries and they are all initialized here. WDYT of having this wrapper-class capture all the static state and having this one capture all the data that we are constructing-on-first-use in the template class?

File: include/envoy/http/header_map.h [hunk 5]
```diff
@@ -637,12 +631,90 @@ class HeaderMap {
...
+  const typename CustomInlineHeaderRegistry<T>::Handle handle_;
+};
+
 /**
  * Typed derived classes for all header map types.
  */
 
 // Base class for both request and response headers.
-class RequestOrResponseHeaderMap : public virtual HeaderMap {
+class RequestOrResponseHeaderMap : public HeaderMap {
```
Comment (jmarantz) [inline include/envoy/http/header_map.h:717]: the de-virtualization of this inheritance is great; but I wonder if you could break this into a separate PR that could be merged without controversy or extensive benchmarking, thus simplifying this one. WDYT?

File: source/extensions/filters/http/cors/cors_filter.cc (hunk 1 above)
Comment (jmarantz) [reply source/extensions/filters/http/cors/cors_filter.cc:29]: agreed this is nice, but it looks like this could be a static-init fiasco. Could this be done in a construct-on-first-use wrapper for each filter?

Or do we need the semantics to be that any filter that is linked in gets its custom headers registered before any packets flow?

File: include/envoy/http/header_map.h [hunk 6]
```diff
@@ -637,12 +631,90 @@ class HeaderMap {
...
+  // Phantom is used here to force the compiler to verify that handles are not mixed between
+  // concrete header map types.
+  using Handle = Phantom<RegistrationMap::const_iterator, T>;
+
+  /**
+   * Register an inline header and return a handle for use in inline header calls. Must be called
+   * prior to finalize().
+   */
+  static Handle registerInlineHeader(const LowerCaseString& header_name) {
+    static size_t inline_header_index = 0;
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply include/envoy/http/header_map.h:652]: > should we wrap this in a construct-on-first-use atomic? IDK if extensions might be brought in from xDS on worker threads and potentially race.

I think for a primitive type it doesn't matter, but I can do it for consistency if you prefer. (Note that this is all pre-main so single threaded.)

> Also having a static declared inline in a templated class declaration makes me unsure how many they'll be. I guess one per template class instantiation? Is that what you wan t?

Right, one per template class, which is on purpose. Let's discuss this more in the other comment thread below.

File: include/envoy/http/header_map.h (hunk 3 above)
Comment (mattklein123) [reply include/envoy/http/header_map.h:654]: The accessor doesn't assert finalized, it just gets the mutable variable.

File: include/envoy/http/header_map.h (hunk 5 above)
Comment (mattklein123) [reply include/envoy/http/header_map.h:717]: Yes I can likely split this into a different PR, but I would rather decide on whether we want to go in this direction or not first, since I'm not sure if it's worth it by itself (given that it introduces it's own templating complexity.)

File: include/envoy/http/header_map.h (hunk 4 above)
Comment (mattklein123) [reply include/envoy/http/header_map.h:807]: > WDYT of having this wrapper-class capture all the static state and having this one capture all the data that we are constructing-on-first-use in the template class?

Yes I think I can likely collapse all of the public access into a single class. It's possible I can even have the bifurcation be using an enum vs. the interface, which would make it even more clear, but I need to see about that, as there are other follow on template implications. If we decide to pursue this direction I can probably simplify this.

File: source/extensions/filters/http/cors/cors_filter.cc (hunk 1 above)
Comment (mattklein123) [reply source/extensions/filters/http/cors/cors_filter.cc:29]: > Or do we need the semantics to be that any filter that is linked in gets its custom headers registered before any packets flow?

Yes, it has to work this way, similar to how extensions are registered. I don't think in practice there is any fiasco as all data is itself accessed using "fiasco proof" paradigms.

With that said, I *think* it might be possible to do something like allow extensions to register headers when they are statically registered somehow. I'm not sure if that is much better though. Either way I think this is something that could be potentially improved later if we go this direction, and later add the ability to register custom headers using CLI/bootstrap which I'm pretty sure is possible.

**Comment by mattklein123** [comment]
@jmarantz thanks for taking an initial look. This week let's discuss whether we want to pursue this direction or not. Overall, I think I'm in favor of proceeding, as it cleans up a bunch of stuff and solves an entire class of perf problems with the current impl, but I'm not wedded to it. If there is enough support I can start working on some benchmarks to further prove it out before finalizing the change.

**Comment by jmarantz** [comment]
@htuch fyi

File: source/extensions/filters/http/cors/cors_filter.cc [hunk 7]
```diff
@@ -59,31 +74,32 @@ Http::FilterHeadersStatus CorsFilter::decodeHeaders(Http::RequestHeaderMap& head
...
       {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::OK))}})};
 
-  response_headers->setAccessControlAllowOrigin(origin_->value().getStringView());
+  response_headers->setInline(access_control_allow_origin.handle(),
+                              origin_->value().getStringView());
 
   if (allowCredentials()) {
-    response_headers->setReferenceAccessControlAllowCredentials(
-        Http::Headers::get().CORSValues.True);
+    response_headers->setReferenceInline(access_control_allow_credentials.handle(),
```
Comment (htuch) [inline source/extensions/filters/http/cors/cors_filter.cc:89]: This looks really pretty usable. How far does this take us towards being able to completely eliminate non-O(1) headers? 

Presumably there will always be some places we need to deal with dynamic headers, e.g. cluster header routing and the like. Are these going to be relatively few?

I still remain curious on whether the most basic flat hash map with ordering support in the value would be competitive. Maybe with an additional map for handles similar to what exists here to improve performance for the well-known header case.

Comment (mattklein123) [reply source/extensions/filters/http/cors/cors_filter.cc:89]: > Presumably there will always be some places we need to deal with dynamic headers, e.g. cluster header routing and the like. Are these going to be relatively few?

It really depends on the use case. For people that are dealing with a lot of dynamic config, especially that they don't control, it may not be reasonable to pre-define all of the "fast" headers they want. This would either necessitate a different implementation entirely, or an escape hatch where we build a map on-demand like @jmarantz prototyped.

> I still remain curious on whether the most basic flat hash map with ordering support in the value would be competitive. Maybe with an additional map for handles similar to what exists here to improve performance for the well-known header case.

I don't know. As we have discussed a bunch of times there are a lot of trade-offs here. The reason this change is appealing to me is a) it cleans up a bunch of stuff and b) it should be better in all cases than what we have now (pending benchmark verification). Later, we can get rid of it all, have pluggable implementations, add a dynamic hash creation setup, etc.

Comment (htuch) [reply source/extensions/filters/http/cors/cors_filter.cc:89]: I'm good with this PR as a strict improvement for the existing implementation for reasons (a) and (b). I think this thread could be used for @adisuissa to track for future directions.

Comment (mattklein123) [reply source/extensions/filters/http/cors/cors_filter.cc:89]: OK, @jmarantz @asraa @adisuissa any objections? If not I will start the cleanup and add some benchmarks, etc.

Comment (asraa) [reply source/extensions/filters/http/cors/cors_filter.cc:89]: This looks great from the filter's perspective -- no objections! I'm happy that some calls for `get()` in filters can now be replaced with something O(1).

A naive question though, basically all the compiled extensions will have their O(1) headers registered, even if at run-time it may not be used -- is that correct?

Comment (mattklein123) [reply source/extensions/filters/http/cors/cors_filter.cc:89]: > A naive question though, basically all the compiled extensions will have their O(1) headers registered, even if at run-time it may not be used -- is that correct?

Yes, basically every registered header will cost 8 bytes in the header map (only the specific map it's registered to) + the time to memset it to 0, whether it's used or not. It's identical to the current system but fully customizable. This means the default map when I'm done will exclude a bunch of headers only used by extensions.

**Comment by adisuissa** [comment]
SGTM.
Really interested in seeing how it compares with a map (performance wise).
Once it's done, I can merge it with #11255 and start playing with this.

**Comment by jmarantz** [comment]
+1 landing this sgtm.

**Comment by mattklein123** [comment]
I will open this again when ready for review.
